AddToPath re-implementation in install-powershell.ps1

glachancecmaisonneuve opened the pull request:
## PR Summary

This PR is a re-implementation of the -AddToPath switch in tools/install-powershell.ps1, without the side-effects the current implementation has.  The changes only affects windows users.  

## Summarized changes

As currently implemented, the -AddToPath switch has quite a few undesirable side-effects and 1 or 2 issues that prevents it from adding the path to the newly installed powershell to the registry.  Specifically:
- it expands all variables in the path to their full value (%SystemRoot% ==> C:\Windows for example) and saves the new path with all variables expanded.
- it changes the path type from REG_EXPAND_SZ to REG_SZ, disabling all variables (wich are now gone in any event).
- it fails o add the path to the newly installed powershell if the user has a path that has the same base path (it won't add c:\pwsh of you have c:\pwsh6 in your path).
- it will add a duplicate path for path values that have a DirectorySeparatorChar at the end (for example, it will add c:\pwsh6\ in your path even if you have c:\pwsh6).

This PR remedies all of the above by
- strictly adding the new path to the current path value, without changing the current path value in any other way.
- saving the new path with the same REG_EXPAND_SZ/REG_SZ it had before, if appropriate. 
- not adding a duplicate path by understanding that C:\1 and C:\1\ is the same path.

## Non-Breaking change
- The updated script will attempt to add the path to the newly installed pwsh in the user's path.  This is entirely new, as the previous version only attempted to save the path in the machine wide path, and very likely failed most of the time, lacking sufficient rights to do so.  
- This last bit also explains why these issues and side-efects have gone unnoticed until now ... you have to be in an elevated console for the path-rewrite to happen.  Since the script is mostly used while building (via build.psm1), and since most people don't use an elevated console for building, ... I beleive few users were affected.

## An alternate fix (dont do anything for -AddToPath when $WinEnv -eq $true)
- I am not convinced of the pertinence of saving the path to the registry.  It isn't required by the build script (adding the path to the current process is enough to satisfy it's needs).  It depends who the target audience for install-powershell.ps1 is.  If the script is used by the public at large as an easy way to install powershell, then it's pertinent.  If it's really only used by the build script, then I wasted quite a bit of time getting this right, and the -AddToPath switch should not do anything for $WinEnv -eq $true.

glachancecmaisonneuve committed:
AddToPath fix

kvprasoon commented on tools/install-powershell.ps1:
```diff
@@ -38,7 +38,7 @@ $ErrorActionPreference = "Stop"
 $IsLinuxEnv = (Get-Variable -Name "IsLinux" -ErrorAction Ignore) -and $IsLinux
 $IsMacOSEnv = (Get-Variable -Name "IsMacOS" -ErrorAction Ignore) -and $IsMacOS
-$IsWinEnv   = !$IsLinuxEnv -and !$IsMacOSEnv
+$IsWinEnv = !$IsLinuxEnv -and !$IsMacOSEnv
```
You could have retained indentation.
```
$IsMacOSEnv = (Get-Variable -Name "IsMacOS" -ErrorAction Ignore) -and $IsMacOS
$IsWinEnv      = !$IsLinuxEnv -and !$IsMacOSEnv
```

glachancecmaisonneuve replied:
Thank you, don't know how i didnt see that sending this in.

glachancecmaisonneuve committed:
Update install-powershell.ps1

glachancecmaisonneuve commented on tools/install-powershell.ps1:
```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
+            (-not ($Path.StartsWith($env:LOCALAPPDATA)))) {
+            try {
+                [string]$KeyName = 'SYSTEM\CurrentControlSet\Control\Session Manager\Environment\'
+                SavePathInRegistryKey([Microsoft.Win32.Registry]::LocalMachine.OpenSubKey($KeyName), $Path)
```
but.... this would reintroduce the problems this PR is trying to avoid!  may as well close the pr then.

glachancecmaisonneuve replied:
### TLDR: nobody wants to touch this with a 10 foot clown pole.

> You should really file an issue on DotNet as well

I'll write a solid draft of everything I know about this, links and all, if you are willing to send a proposal.

### How the windows api team sees this (as imagined by me)

I like the way the windows api sees things myself.  Their view on the matter (this is all interpretation on my part from what the api does) is that there is no such thing as an _environment variable_ outside of a process.  [The docs clearly state that a call to SetEnvironmentVariable changes the process environment block (PEB) only](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-setenvironmentvariable).  They offer completely different methods for saving values in the registry.

### Design issues with SetEnvironmentVariable

The dotnet team tried to _fuzz_ the distinction between a _variable-living-inside-a-process_ and a _setting-whose-value-is-set-in-the-registry_.  The ["target" parameter](https://referencesource.microsoft.com/mscorlib/a.html#821defd26e0b02d7) (which does not exist in the windows api embodies a perfect example of a single responsibility principle violation.

It's possible values are process, user and machine.  The first one implies changes in memory while the other two are clearly about changing values in the registry.   

### Implementation details in SetEnvironmentVariable
For the 2 targets that end up targeting the registry, i.e. USER and MACHINE,
[SetEnvironmentVariable relies on Microsoft.Win32.Registry](https://referencesource.microsoft.com/mscorlib/a.html#33fd9bc66f5473b7) class to implement registry persistence.  

Because SetEnvironmentVariable doesn't provide an overload with a RegistryValueKind tyoe parameter however, it defaults to saving RegistryValueKind.REG_SZ, disabling expansion on any %VARS% in the path.  Cutting out the middle man (SetEnvironmentVariable) and and relying on Microsoft.Win32.Registry is the only option available.

### What would need to happen for a fix in Environment
- All things considered, the easiest path seems to be to add an overload to SetEnvironmentVariable  that would accept a RegistryValueKind as a parameter.
- Adding a RegistryValueKind parameter will surely bring requests to support REG_BINARY, REG_MULTI_SZ and others as well, meaning more overloads (one with a value parameter that is of type Byte[], that will persist as a REG_BINARY, one with a value parameter of type Int that will persist as a REG_DWORD and implement auto-detection of REG_MULTILINE_SZ).

### Why I think any proposal coming from outside Microsoft won't be fixed?
The payoff is small (set the value of two miserable registry values). and the other reason isn't very compelling = we're doing this so users don't have to think about the difference between a registry value and variables living in a PEB.

### Why I think you might have a chance
In theory the golden rule that asks everyone to **always assume competence** from the person that wrote the PR or the bug report or the code review is correct.  A few years of badly written PRs, angry and/or  disrespectful bug reports can't help but to quickly chip away at this golden rule idea however.  A PR coming from a colleague has a chance to be taken seriously.

glachancecmaisonneuve replied:
(no changes to code)

glachancecmaisonneuve commented on tools/install-powershell.ps1:
```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
+    if (-not (IsPathInRegistryPath($Path))) {
+        try {
+            [string]$KeyName = 'Environment'
+            SavePathInRegistryKey([Microsoft.Win32.Registry]::CurrentUser.OpenSubKey($KeyName), $Path)
```
but.... this would reintroduce the problems this PR is trying to avoid!  may as well close the pr then.

glachancecmaisonneuve replied:
(no changes to code)

TravisEz13 committed:
Update tools/install-powershell.ps1

Co-Authored-By: glachancecmaisonneuve <glachance@cmaisonneuve.qc.ca>

TravisEz13 committed:
Update tools/install-powershell.ps1

Co-Authored-By: glachancecmaisonneuve <glachance@cmaisonneuve.qc.ca>

glachancecmaisonneuve commented on tools/install-powershell.ps1:
```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
+    return $False
+}
+
+Function SavePathInRegistryKey([Microsoft.Win32.RegistryKey] $Key, [string] $Path) {
```
You are a much better man than I.  I tend to avoid using a "-" in any function that wont be public to avoid the discoverability linting message.

glachancecmaisonneuve replied:
Your suggestion on using aprooved verb for all functions, public or not, was implemented in the latest psuh.  I chose to go with Add-Path (~proposed), Test-PathInRegistryPath (===proposed) and Update-Path (===proposed).

TravisEz13 committed:
Update tools/install-powershell.ps1

Co-Authored-By: glachancecmaisonneuve <glachance@cmaisonneuve.qc.ca>

TravisEz13 committed:
Update tools/install-powershell.ps1

Co-Authored-By: glachancecmaisonneuve <glachance@cmaisonneuve.qc.ca>

glachancecmaisonneuve committed:
Code-Review changes

glachancecmaisonneuve commented on tools/install-powershell.ps1:
```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
     }
 }
 
+Function IsPathInRegistryPath([string] $Path) {
```
done

glachancecmaisonneuve commented on tools/install-powershell.ps1:
```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
+    #Possible gotcha: don't use Get-ItemPropertyValue (see https://github.com/PowerShell/PowerShell/issues/5906)
+    #Possible gotcha: coerce the (possibly null) property to string before calling ExpandEnvironmentVariables
+    [string] $HKCURegKeyPath = Get-ItemProperty -Path 'HKCU:\Environment' -Name 'PATH' -ErrorAction SilentlyContinue
+    $InstalledPaths += @([System.Environment]::ExpandEnvironmentVariables(($HKCURegKeyPath)).Split(';'))
```
> You should not expand any part of the path you did not add.

This function checks if the path is already in the users path.  **It need to expand everything.**  

If $Destination is "%APPDATA%\1" and the PATH already has "%USERPROFILE%\Appdata\Roaming\1" then I don't want to add $Destination to the Path.... it's already in there.

In any event, the proposed change ([System.Environment]::GetEnvironmentVariable('PATH','User')) expands all variables so this change was commited in https://github.com/PowerShell/PowerShell/pull/8081/commits/2fe05aa5d07c220ac077c1fc52ef4231aad66286

glachancecmaisonneuve commented on tools/install-powershell.ps1:
```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
+    #Possible gotcha: coerce the (possibly null) property to string before calling ExpandEnvironmentVariables
+    [string] $HKCURegKeyPath = Get-ItemProperty -Path 'HKCU:\Environment' -Name 'PATH' -ErrorAction SilentlyContinue
+    $InstalledPaths += @([System.Environment]::ExpandEnvironmentVariables(($HKCURegKeyPath)).Split(';'))
+    [string] $HKLMRegKeyPath = Get-ItemProperty -Path 'HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\Environment' -Name 'PATH' -ErrorAction SilentlyContinue
```
done in https://github.com/PowerShell/PowerShell/pull/8081/commits/2fe05aa5d07c220ac077c1fc52ef4231aad66286

glachancecmaisonneuve commented:
Sorry it took so long to implemented the simple changes you were asking for.  I am not accustomed to this form of review were code changes can be accepted... and forgot that I could just push the requested changes by myself, just as usual.

It must have seemed like I was either being uncommonly stubborn/dimwitted/uncooperative.

Thank you for taking the time and your efforts towards integrating this code review.

**Reminder:** There is one proposed change that has not been implement, i.e. using [Environment]::SaveEnvironmentVariable to write to the registry as it lacks the ability to assign a RegistryValueKind.

glachancecmaisonneuve commented on tools/install-powershell.ps1:
```diff
@@ -79,6 +79,68 @@ Function Remove-Destination([string] $Destination) {
+    $Key.SetValue("PATH", $NewPathValue, $PathValueKind)
+}
+
+Function AddPathToPathWinEnv([string]$Path) {
```
[done](https://github.com/PowerShell/PowerShell/pull/8081/commits/2fe05aa5d07c220ac077c1fc52ef4231aad66286)

glachancecmaisonneuve commented:
I just found out that the Set-ItemProperty cmdlet honnors the registryvaluekind when provided [ref](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_registry_provider?view=powershell-6#type-microsoftwin32registryvaluekind).

Would you have preferred that over a call to [Microsoft.Win32.RegistryKey]::SetValue?

glachancecmaisonneuve commented:
Sorry it's the end of the term and I haven't had much time at all.  I'll implement the requested changes as soon as i'm able

glachancecmaisonneuve committed:
It occured to me while trying to document the functions that:
- Test-PathInRegistryPath was in fact a validator
- Update-Path should not be a function but in the script's body
- Add-PathToSettings was the only real function

Added:
- parameter validation
- function documentation

Other:
- formatting changes as requested

TravisEz13 committed:
Merge branch 'master' into install-powershell-addtopath-fix

glachancecmaisonneuve committed:
A few style changes and bug fix with invalid destination
